rfc/src/rfcs/0010-rust-ub-checks.md

```diff
@@ -0,0 +1,175 @@
+However, Kani should provide a fine-grained control for users to choose which checks to disable.
+
+For every class of UB checks to be added, we shall add a command line argument to disable the check for all harnesses,
+as well as an attribute that would allow users to opt-out a check for specific harnesses.
+
+Command line arguments should follow the existing name schema:
+`--no-[CHECK-NAME]-checks` where the CHECK-NAME is kebab case. There is no need to add `--[CHECK-NAME]-checks` since
+they are enabled by default.
+
+We propose to add the following attribute:
+
+```rust
+#[kani::proof]
+#[kani::ignore(CHECK_NAMES)]
+fn harness() {}
+```
+
+Where `CHECK_NAMES` is a list of check category names in snake case, such as `memory_safety`, `valid_values`.
+
+For checks that are unstable, Kani should support both opt-out mechanisms.
+These checks shall be included by default if their respective unstable flag are passed, i.e., `-Z[CHECK-NAME]`
+
+## Software Design
+
+*To be filled in the next iteration*
+
+## Rationale and alternatives
+
+Kani already has undefined behavior checks, some are implemented on the CBMC side, such as access out bounds, and
+arithmetic overflow.
+While others are implemented on Kani's side, such as intrinsics validation checks, and some are a
+result of instrumentation to the Rust standard library.
+
+The behavior of these checks varies according to how they are added.
+The ones on CBMC side do not interrupt the analysis of a path, while the ones on Kani / Rust std side will.
+
+For example, memory bound checks are implemented in CBMC, and the following harness today will trigger 2 different
+failures:
+
+```rust
+#[kani::proof]
+pub fn is_zero() {
+    let var = [0u32; 4];
+    let ptr = &var as *const u32;
+    let idx: usize = 4;
+    assert_eq!(unsafe { ptr.add(idx).read() }, 0);
+}
+```
+
+The failed checks are:
+
+```plaintext
+Failed Checks: assertion failed: unsafe { ptr.add(idx).read() } == 0
+Failed Checks: dereference failure: pointer outside object bounds
+```
+
+While the following harness only fails the standard library check:
+
+```rust
+#[kani::proof]
+pub fn is_zero() {
+    let var = [0; 4];
+    let idx: usize = 4;
+    assert_eq!(unsafe { *var.get_unchecked(idx) }, 0);
+}
+```
+
+Failing due to an assumption check inside `get_unchecked` function.
+
+```plaintext
+Failed Checks: assumption failed
+```
+
+We would like to keep taking advantage of existing instrumentation to maximize the UB check coverage.
+The mechanism used to build the check, however, shouldn't influence in the user experience.
+
+### Opt-out vs opt-in
+
+The default behavior matches the one that we believe is the recommended for most users due to soundness.
+In some cases, users may need to disable some checks for reasons such as spurious CEX or proof scalability.
+Thus, we suggest that every category of checks should have an opt-out mechanism that can either be global or local;
+i.e.: via command argument or harness attribute.
```

Comment:
Since some of our checks are over-approximating, we may also need to introduce code-level control at some point, e.g. to enable/disable the checks for a few lines of code.

Reply:
Indeed. It might be hard for dependencies though.

Reply:
+1

Reply:
Is it OK if I leave this as future work?